drivers/can/can_nxp_s32_canxl.c

```diff
@@ -0,0 +1,1020 @@
+	CAN_NXP_S32_CTRL_CALLBACK(n)							\
+	CAN_NXP_S32_ERR_CALLBACK(n)							\
+	CAN_NXP_S32_IRQ_CONFIG(n)							\
+	CAN_NXP_S32_IRQ_ENABLE(n)							\
+	CAN_NXP_S32_IRQ_DISABLE(n)							\
+	PINCTRL_DT_DEFINE(CAN_NXP_S32_NODE(n));						\
+	Canexcel_Ip_ConfigType can_nxp_s32_default_config##n = {			\
+		.rx_mbdesc = (uint8)CAN_NXP_S32_MAX_RX,					\
+		.tx_mbdesc = (uint8)CAN_NXP_S32_MAX_TX,					\
+		.CanxlMode = CANEXCEL_LOOPBACK_MODE,					\
+		.fd_enable = (boolean)CAN_NXP_S32_FD_MODE,				\
+		.bitRateSwitch = (boolean)CAN_NXP_S32_BRS,				\
+		.ctrlOptions = (uint32)CANXL_IP_BUSOFF_RECOVERY_U32,			\
+		.Callback = nxp_s32_can_##n##_ctrl_callback,				\
+		.ErrorCallback = nxp_s32_can_##n##_err_callback				\
+	};										\
+	__nocache Canexcel_Ip_StateType can_nxp_s32_state##n;				\
+	__nocache Canexcel_TxFdMsgType tx_msg##n[CAN_NXP_S32_MAX_TX];			\
+	__nocache Canexcel_RxFdMsg rx_msg_##n[CAN_NXP_S32_MAX_RX];			\
+	static struct can_nxp_s32_data can_nxp_s32_data_##n = {				\
+		.can_state = (Canexcel_Ip_StateType *)&can_nxp_s32_state##n,		\
+		.tx_msg = tx_msg##n,							\
+		.rx_msg = rx_msg_##n,							\
+	};										\
+	static struct can_nxp_s32_config can_nxp_s32_config_##n = {			\
+		.base_sic = (CANXL_SIC_Type *)						\
+				DT_REG_ADDR_BY_NAME(CAN_NXP_S32_NODE(n), sic),		\
+		.base_grp_ctrl = (CANXL_GRP_CONTROL_Type *)				\
+				DT_REG_ADDR_BY_NAME(CAN_NXP_S32_NODE(n), grp_ctrl),	\
+		.base_dsc_ctrl = (CANXL_DSC_CONTROL_Type *)				\
+				DT_REG_ADDR_BY_NAME(CAN_NXP_S32_NODE(n), dsc_ctrl),	\
+		.instance = n,								\
+		.clock_can = DT_PROP(CAN_NXP_S32_NODE(n), clock_frequency),		\
+		.bitrate = DT_PROP(CAN_NXP_S32_NODE(n), bus_speed),			\
+		.sjw = DT_PROP(CAN_NXP_S32_NODE(n), sjw),				\
+		.prop_seg = DT_PROP_OR(CAN_NXP_S32_NODE(n), prop_seg, 0),		\
+		.phase_seg1 = DT_PROP_OR(CAN_NXP_S32_NODE(n), phase_seg1, 0),		\
+		.phase_seg2 = DT_PROP_OR(CAN_NXP_S32_NODE(n), phase_seg2, 0),		\
+		.sample_point = DT_PROP_OR(CAN_NXP_S32_NODE(n), sample_point, 0),	\
+		CAN_NXP_S32_TIMING_DATA_CONFIG(n)					\
+		.max_bitrate = DT_CAN_TRANSCEIVER_MAX_BITRATE(CAN_NXP_S32_NODE(n),	\
+							CAN_NXP_S32_MAX_BITRATE),	\
+		.phy = DEVICE_DT_GET_OR_NULL(DT_PHANDLE(CAN_NXP_S32_NODE(n), phys)),	\
+		.pin_cfg = PINCTRL_DT_DEV_CONFIG_GET(CAN_NXP_S32_NODE(n)),		\
+		.can_cfg = (Canexcel_Ip_ConfigType *)&can_nxp_s32_default_config##n,	\
+		.irq_config_func = can_irq_config_##n,					\
+		.irq_enable_func = can_irq_enable_##n,					\
+		.irq_disable_func = can_irq_disable_##n					\
+	};										\
+	static int can_nxp_s32_##n##_init(const struct device *dev)			\
+	{										\
+		return can_nxp_s32_init(dev);						\
+	}										\
+	DEVICE_DT_DEFINE(CAN_NXP_S32_NODE(n),						\
+			&can_nxp_s32_##n##_init,					\
+			NULL,								\
+			&can_nxp_s32_data_##n,						\
+			&can_nxp_s32_config_##n,					\
+			POST_KERNEL,							\
+			CONFIG_CAN_INIT_PRIORITY,					\
+			&can_nxp_s32_driver_api);
+
+#if DT_NODE_HAS_STATUS(CAN_NXP_S32_NODE(0), okay)
+CAN_NXP_S32_INIT_DEVICE(0)
```

Comment:
Nit: Using instances like this is HAL specific and does not follow the Zephyr design principles. You should at least check for the right compatible string and not just rely on the nodelabel.

Reply:
I agree. We are planning to address this issue for all S32 peripheral drivers added so far. The limitation is that the HAL is designed in such a way that needs to know the peripheral hw instance index, sometimes at build time. In previous reviews we had the blessing to merge as it is using nodelabel, so I would suggest keeping this topic to be addressed as an improvement for the whole S32 port in a separate PR, if possible.

Reply:
Sounds good. Not a blocker from my side, just commenting on it for the record.